native: timer: synchronize HWTIMER_SPIN_BARRIER with timer min resolu

benoit-canet opened the pull request:
…tion

This keeps the test/hwtimer_wait test duration correct.

PeterKietzmann commented on tests/hwtimer_wait/main.c:
```diff
@@ -30,7 +30,7 @@ int main(void)
 
     long iterations = 10000;
-    long start_duration = (HWTIMER_SPIN_BARRIER + 1) << 5;
+    long start_duration = (20 + 1) << 5;
```
IIRC the idea was to start the test at the lower threshold of `HWTIMER_SPIN_BARRIER` so the spin function is not involved. What do you think about leaving this line as it was but decreasing the number of iterations so the test won't take so long? Or did I intercept something you discussed in the latest timer-related PRs?

PeterKietzmann commented:
Maybe I'm the first one saying this: "Strider" is happy :-) . ACK when Travis is also.

benoit-canet commented:
Hmm it was breaking rpl_udp I fixed it.

kaspar030 commented:
My gut feeling is that 200 is a lot for spinning, and we should be sure that native can't do better before enforcing it.

kaspar030 commented:
@benoit-canet How did it break rpl? And, please don't squash while the review is in progress, it sometimes makes it hard to follow changes.

benoit-canet commented:
I added this define in native_internal.h so the ng stack was double including some stuff and compilation failed.

I tried 10 20 50 100 and 200 as values. And 200 was the first value to work reliably.

I added this minimal value because I saw the torture hwtimer_wait test script fail simultaneously on
two of my linux boxes. I guess is related to clock adjustement by ntp or something like that.
That would explain that lower values still fails from time to time.

benoit-canet commented:
Hmm I am retesting hwtimer_wait before this patch with a value of 20 and it works.
I guess the time drift compensation was just trampoline rewrite compensation :)

PeterKietzmann commented:
@kaspar030 may I assign you?

PeterKietzmann commented:
@kaspar030 I'll just do it. If you don't agree, reassign me...

kaspar030 commented:
@PeterKietzmann no problemo

@benoit-canet hmm, I feel like we're more trying than thinking. :)

Let me try to summarize :

- hwtimer_spin_barrier was introduced because the overhead of setting a timer, letting it trigger, return to the right context is not zero. Also, while setting a timer it's trigger point might pass, causing an underflow. So below a certain timer period length, it makes more sense to just spin, or is even needed.
hwtimer will go soon, so not optimizing it's spin value is IMHO fine.

- native's timer_set_absolute has some overhead, too, making underflows possible:
1. code calls timer_set_absolute at t + 0, setting t + 1
2.  timer_set_absolute calls timer_read(), gets t==2, sets itimer to  1-2=0xFFFFFFFE

These cases should now be caught by the delta check, but we need to find out if "200" is a generally acceptable value and why it fluctates. Might be the scheduling on the host system that influences this.

Maybe we should write a test?

- when developing wtimer, I tried to make sure that timers trigger at the expected trigger time +- 1 tick. Somehow for very short periods, native/timer (itimer) fluctuated too much, so to get reliable results, I introduced NATIVE_TIMER_MIN_RES. We might need to reevaluate the need for this.

kaspar030 commented:
(I'm not merging this as quickly as the changes yesterday because I think stability is premium, while accuracy can get a little thought time)

benoit-canet commented:
@kaspar030 : I don't think this patchset impact the stability: it just change the place where the 200 constant is defined.

I'll try to dig in order to find why 200.

kaspar030 commented:
> it just change the place where the 200 constant is defined.

Well, it raises the hwtimer spin barrier from 20 to 200, right?

benoit-canet commented:
yes

LudwigOrtmann commented on tests/hwtimer_spin/main.c:
```diff
@@ -32,7 +32,7 @@ int main(void)
     puts("If you do not see that message, something went wrong.\n");
 
-    for (unsigned long r = 10000; r > 0; r--) {
+    for (unsigned long r = 1000; r > 0; r--) {
```
is this unrelated?

benoit-canet commented:
The test was 10 time longer because the spin barrier is ten time higher.

benoit-canet commented:
@kaspar030 In fact it's me who edited your original perif/timer.c patchset and added the  NATIVE_TIMER_MIN_RES constant hopping there would be a way to dual Signed-Of-By this particular patch like it's done in QEMU when two people work on the same patch.

So I am pretty sure that the code doesn't work with low value for  NATIVE_TIMER_MIN_RES.
I tested it again today with 20 and after a few hours the test failed.

benoit-canet commented:
@kaspar030 What bug me is that I don't understand _why_ the code doesn't work without this minimal resolution. I need to sort this out.

benoit-canet commented:
I failed to reproduce the stall with a loop doing hwtimer_wait(1) without the NATIVE_TIMER_MIN_RES.

So I think these freeze occurs on an exceptional system event.

kaspar030 commented:
> I failed to reproduce the stall with a loop doing hwtimer_wait(1) without the NATIVE_TIMER_MIN_RES.
> 
> So I think these freeze occurs on an exceptional system event.

Maybe we fixed this by catching underflows in timer_set_absolute() and
also by not disabling the timer when doing timer_set(0)?

benoit-canet commented:
I checked this. it still fail from time to time with master and a value of 1.

benoit-canet commented:
@kaspar030 Still faill with master and 20

benoit-canet commented:
One possible plain simple explanation is that the minimal timer value gives to the setitimer syscall the time to return before the signal handler is fired since in syscall signals are postponed.

kaspar030 commented:
@benoit-canet that sounds likely here. So we do need a spin barrier in periph/timer.c to get this stable.

benoit-canet commented:
@kaspar030 what do we do about this pr ? The code seems to compile ok here yet travis fails on cortex uc .

kaspar030 commented:
@benoit-canet It missed the "Ready for CI" label...

benoit-canet commented:
@kaspar030 hmm cortex still faill and no logs are provided by travis

benoit-canet commented:
@kaspar030 : I tried to test the linux timer in a simple test program doing something similar to native.

The result is here:
http://paste.debian.net/283827/

I ran it with: ./a.out &> l

Then done
cut -f 1 -d ' ' l|uniq

This gives:
http://paste.debian.net/283829/

where you can see thing like:

1437305556
1437305557
1437305563
1437305564

where the code is stalled during six seconds.

It's rather strange.

I'll try to run it a few hours to see if it doesn't stall.

benoit-canet commented:
@kaspar030 I made my test program more close to what's happening for real in native and got the explication of the drama.

http://paste.debian.net/283830/

So if the timer fire before the iddle loop enter in pause() native will be stuck.

benoit-canet commented:
I think I finally nailed it with a clean solution. I will test it for the night.

benoit-canet commented:
@kaspar030: I tried to get rid of NATIVE_TIMER_MIN_RES without success. It's not a clean solution but I think we need this kludge for now.

benoit-canet commented:
@kaspar030 given how the continuous integration test fails this pr is correct and the CI machine doesn't support ipv6:

Extract of the log:
W: Failed to fetch http://ftp.uk.debian.org/emdebian/toolchains/dists/wheezy/main/binary-amd64/Packages 404 Not Found [IP: 2001:1b40:5600:ff80:f8ee::1 80]

W: Failed to fetch http://ftp.uk.debian.org/emdebian/toolchains/dists/wheezy/main/binary-i386/Packages 404 Not Found [IP: 2001:1b40:5600:ff80:f8ee::1 80]

kaspar030 commented:
@benoit-canet I agree.

ACK.
